### DISCUSSION TITLE
Hidden scrollbar doesn't resize with window

### DISCUSSION THREAD

**Discussion by fiddler55** [fltk-rs/fltk-rs#1592] [Q&A]
A window with a flex layout widget has a horizontal scrollbar at the bottom. When the window is resized the scrollbar properly adjusts to the new size. The problem occurs when the scrollbar is hidden and then the window is resized. When the scrollbar is re-shown it appears in it's prior position. When the window is resized again the scrollbar will then snap to the new position and work properly. The following two images demonstrate this behavior. The top image shows the window when it first starts up. The bottom image shows what happens when the scrollbar is hidden, the window is resized, and then the scroll bar is shown.

![ScrollHide1](https://github.com/user-attachments/assets/dfd24aef-401b-487e-95fb-4c9cbde00fed)

![ScrollHide2](https://github.com/user-attachments/assets/b0446b73-243e-4eb7-8dd5-bbfa5c9cd35a)

The follow code can demonstrate the behavior:

```
// Testing the fltk scrollbar with hiding and resizing
// Be sure to have the following line in Cargo.toml after [dependencies]:
// fltk = { version = "1.4.15", features = ["fltk-bundled"] }
use fltk::{prelude::*, window::Window, enums::*, *};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default()
        .with_size(400, 100)
        .with_label("Scrollbar hide test");
    wind.make_resizable(true);

    let mut col = group::Flex::default_fill().column();
    col.set_pad(0);

    let mut mbar: menu::MenuBar = menu::MenuBar::default();
    col.fixed(&mbar, 25);
    mbar.add(
        "&Options/Hide scrollbar",
        Shortcut::None,
        menu::MenuFlag::Toggle,
        menu_cb,
    );

    let mut _text = frame::Frame::default().with_label("Some text");

    let mut _bar = valuator::Scrollbar::default()
        .with_type(valuator::ScrollbarType::Horizontal);
    col.fixed(&_bar, 20);

    col.end();

    wind.end();
    wind.show();

    app.run().unwrap();
}

fn menu_cb(mbar: &mut impl MenuExt) {
    if let Ok(mpath) = mbar.item_pathname(None) {
        if mpath.as_str() == "&Options/Hide scrollbar" {
            let state = mbar.find_item("&Options/Hide scrollbar").unwrap().value();
            let mut scroll = mbar.parent().unwrap().child(2).unwrap();
            if state { scroll.hide(); }
            else     { scroll.show(); }
        }
    }
}

```
Take these steps to show the behavior:

1) Start the program (should look like the top image).
2) Click "Options", then "Hide scrollbar". The scrollbar will disappear, but "Some text" doesn't lower to the middle.
3) With mouse, drag the lower the window's lower-right corner down and to the right just a bit. "Some text" suddenly re-centers and then stays in the middle.
4) Click "Options", then "Hide scrollbar" (thereby toggling it off). The scrollbar appears in its prior position, and "Some text" stays in the middle.
5) Drag the window's lower-right corner again. The scrollbar will suddenly jump to its proper position and "Some text" will reposition itself properly.

Remedies? Is there a command I can issue so that everything is repositioned as if the window was resized, but instead the the scrollbar was hiddened/shown? Thanks in advance for any help.

**Comment by MoAlyousef** [comment]
A remedy:
```rust
fn menu_cb(mbar: &mut impl MenuExt) {
    if let Ok(mpath) = mbar.item_pathname(None) {
        if mpath.as_str() == "&Options/Hide scrollbar" {
            let state = mbar.find_item("&Options/Hide scrollbar").unwrap().value();
            let flex = group::Flex::from_dyn_widget(&mbar.parent().unwrap()).unwrap();
            let mut scroll = flex.child(2).unwrap();
            if state { scroll.hide(); }
            else     { scroll.show(); }
            flex.recalc();
            flex.parent().unwrap().set_damage_type(Damage::All);
        }
    }
}
```
A flex deals with hiding a child as if it doesn't exist. Only on resize will it recalculate the size of its children. The modification above forces the flex to recalculate.
The extra call to `flex.parent().unwrap().set_damage_type(Damage::All); // or flex.parent().unwrap().redraw();` is because by default Flex has a NoBox FrameType, so redrawing children might cause artifacts. You can either give the Flex a FrameType::FlatBox for example in main, or ask it's parent to redraw.

**Comment by fiddler55** [comment]
Thanks Mohammed! I figured there had to be an easy way to do this. I did come up with another approach: setting the horizontal scrollbar's height to be 0 so it doesn't go away, it just can't be seen. Then when it is to show again I set its height back to normal. (I also tell the text frame above the scrollbar to redraw its label ("Some text") so that it appears in the correct position.) This approach works, but yours is more elegant...
